### PR TITLE
Change audio component invalidation logic to use flag instead of delegate

### PR DESCRIPTION
**Pull request by peppy**
Fixes https://github.com/ppy/osu/issues/10727 (the focus-related portion).

### CODE REVIEW

File: osu.Framework/Audio/AudioComponent.cs [hunk 1]
```diff
@@ -101,8 +105,13 @@ public void Update()
                 task.RunSynchronously();
 
             if (!IsDisposed)
+            {
                 UpdateState();
 
+                if (Looping && HasCompleted)
```
Comment (peppy) [inline osu.Framework/Audio/AudioComponent.cs:111]: Had to move this out to allow calling `UpdateState` in start calls.

File: osu.Framework/Audio/Track/TrackBass.cs [hunk 2]
```diff
@@ -263,6 +263,9 @@ public override void Start()
 
         private bool startInternal()
         {
+            // ensure state is correct before starting.
+            UpdateState();
```
Comment (peppy) [inline osu.Framework/Audio/Track/TrackBass.cs:267]: without this the frequency handling test will fail. it's probably best we call this so there's never any delay between a track starting and its attributes being updated, though.

**Comment by bdach** [comment]
Tested this for a few minutes, seems to resolve the issue at least. The changes look *mostly* safe to me - operation order is changed somewhat (i.e. in `TrackVirtual`) but I can't see it having a major impact on execution/semantics. Definitely not merging on my own, though.

**Review by bdach — APPROVED** [review]

**Comment by smoogipoo** [comment]
What part of this fixes the focus issue? Somehow delegates not being cleared out?

**Comment by bdach** [comment]
Yep, tasks being enqueued but never dequeued.

**Comment by smoogipoo** [comment]
And why do the tasks never get dequeued? Isn't the audio thread running much faster than the update thread when unfocused (1000fps compared to the 30fps or whatever the update thread is limited to)?

**Comment by peppy** [comment]
Well there's two parts: the actual action overhead (up to millions of actions being created per second, several megabytes). That is what I'm aiming to solve here.

Anything else needs potential further investigation, like why they are not being processed (somehow not in the alive audio hierarchy? not sure).

**Comment by smoogipoo** [comment]
Ok. I'll merge this for now then as I don't disagree with the underlying changes.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**